### ISSUE TITLE
2.3.2 fails to build with GCC 8

### ISSUE THREAD

**Issue by sebastic** [ossimlabs/ossim#200]
To test if [Debian Bug #897829](https://bugs.debian.org/897829) was an actual issue, OSSIM 2.3.2 has been built with GCC 8 and unfortuntely it fails to build:
```
[ 91%] Linking CXX executable ../../bin/ossim-adrg-dump
cd /build/ossim-2.3.2/debian/build/apps/ossim-adrg-dump && /usr/bin/cmake -E cmake_link_script CMakeFiles/ossim-adrg-dump.dir/link.txt --verbose=1
/usr/bin/g++-8  -g -O2 -fdebug-prefix-map=/build/ossim-2.3.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -O3 -DNDEBUG  -Wl,-z,relro -Wl,-z,now -rdynamic CMakeFiles/ossim-adrg-dump.dir/ossim-adrg-dump.cpp.o  -o ../../bin/ossim-adrg-dump -Wl,-rpath,/build/ossim-2.3.2/debian/build/lib: ../../lib/libossim.so.1.9.0 -lgeos -lgeos_c -lgeotiff -ljpeg -ltiff -ljsoncpp -ldl -ljsoncpp -ldl -lfreetype -lz -lpthread 
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<double>(double, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<int>(int, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<unsigned short>(unsigned short, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<float>(float, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<unsigned char>(unsigned char, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<unsigned int>(unsigned int, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<short>(short, ossimImageData*, unsigned int)'
collect2: error: ld returned 1 exit status
make[3]: *** [apps/ossim-adrg-dump/CMakeFiles/ossim-adrg-dump.dir/build.make:99: bin/ossim-adrg-dump] Error 1
make[3]: Leaving directory '/build/ossim-2.3.2/debian/build'
make[2]: *** [CMakeFiles/Makefile2:303: apps/ossim-adrg-dump/CMakeFiles/ossim-adrg-dump.dir/all] Error 2
make[2]: *** Waiting for unfinished jobs....
[ 91%] Linking CXX executable ../../bin/ossim-autreg
cd /build/ossim-2.3.2/debian/build/apps/ossim-autreg && /usr/bin/cmake -E cmake_link_script CMakeFiles/ossim-autreg.dir/link.txt --verbose=1
/usr/bin/g++-8  -g -O2 -fdebug-prefix-map=/build/ossim-2.3.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -O3 -DNDEBUG  -Wl,-z,relro -Wl,-z,now -rdynamic CMakeFiles/ossim-autreg.dir/ossim-autreg.cpp.o  -o ../../bin/ossim-autreg -Wl,-rpath,/build/ossim-2.3.2/debian/build/lib: ../../lib/libossim.so.1.9.0 -lgeos -lgeos_c -lgeotiff -ljpeg -ltiff -ljsoncpp -ldl -ljsoncpp -ldl -lfreetype -lz -lpthread 
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<double>(double, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<int>(int, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<unsigned short>(unsigned short, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<float>(float, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<unsigned char>(unsigned char, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<unsigned int>(unsigned int, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<short>(short, ossimImageData*, unsigned int)'
collect2: error: ld returned 1 exit status
make[3]: *** [apps/ossim-autreg/CMakeFiles/ossim-autreg.dir/build.make:99: bin/ossim-autreg] Error 1
make[3]: Leaving directory '/build/ossim-2.3.2/debian/build'
make[2]: *** [CMakeFiles/Makefile2:413: apps/ossim-autreg/CMakeFiles/ossim-autreg.dir/all] Error 2
[ 91%] Linking CXX executable ../../bin/ossim-applanix2ogeom
cd /build/ossim-2.3.2/debian/build/apps/ossim-applanix2ogeom && /usr/bin/cmake -E cmake_link_script CMakeFiles/ossim-applanix2ogeom.dir/link.txt --verbose=1
/usr/bin/g++-8  -g -O2 -fdebug-prefix-map=/build/ossim-2.3.2=. -fstack-protector-strong -Wformat -Werror=format-security -Wdate-time -D_FORTIFY_SOURCE=2 -O3 -DNDEBUG  -Wl,-z,relro -Wl,-z,now -rdynamic CMakeFiles/ossim-applanix2ogeom.dir/ossim-applanix2ogeom.cpp.o  -o ../../bin/ossim-applanix2ogeom -Wl,-rpath,/build/ossim-2.3.2/debian/build/lib: ../../lib/libossim.so.1.9.0 -lgeos -lgeos_c -lgeotiff -ljpeg -ltiff -ljsoncpp -ldl -ljsoncpp -ldl -lfreetype -lz -lpthread 
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<double>(double, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<int>(int, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<unsigned short>(unsigned short, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<float>(float, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<unsigned char>(unsigned char, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<unsigned int>(unsigned int, ossimImageData*, unsigned int)'
../../lib/libossim.so.1.9.0: undefined reference to `void ossimPixelFlipper::flipPixels<short>(short, ossimImageData*, unsigned int)'
collect2: error: ld returned 1 exit status
```

To reproduce the issue, build OSSIM in a Debian unstable environment with the `gcc-8` & `g++-8` packages installed, and use these with:
```
export CC=gcc-8
export CXX=g++-8
```

**Comment by gpotts** [comment]
Hello Bas:

Is this only for pixelflip or is this for any application linking to OSSIM.


Take care

Garrett Potts

**Comment by gpotts** [comment]
Hello Bas:


Never mind.  It is any app linking to the library.

Let me see if I see anything obvious


Take care

Garrett

**Comment by gpotts** [comment]
Hello Bas:


I think I see the problem.  The templae methods were put into the source an not in the headers.  Surprised this didn’t have problems before

Let me do a quick fix this morning and I will let you know when you can test.


Take care

Garrett

**Comment by omarossim** [comment]
Hello:

Actually I am not sure.  Might need to look for a few.  Seems others do the same thing but seems you are only getting undefined symbols when doing gcc 8 on debian.

Take care

Garrett

**Comment by gpotts** [comment]
Hello Bas:

I saw in the BitMAskWriter that it was calling a template method directly.  That method should not be called directly.  I made the modification but have no way of testing that bitmask writer.  Could you test the build on a debian running gcc 8?



Dave,  do you have a test for bitmask writer?


Take care

Garrett Potts

**Comment by sebastic** [comment]
I've included 545b65169484ad0d5237bf5c6d0279558538b62b as a patch in the Debian package, and that fixes the build failure with GCC 8. Thank!

**Comment by gpotts** [comment]
Hello Bas:

Good deal!!

Thank you for the verification

Take care

Garrett
